{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "from operator import add\n",
    "from math import log\n",
    "import csv\n",
    "import pickle\n",
    "import sys\n",
    "from collections import Counter\n",
    "import re\n",
    "import glob, os\n",
    "\n",
    "import ntpath\n",
    "import functools\n",
    "import itertools\n",
    "\n",
    "from time import time\n",
    "import argparse\n",
    "from PRS_run import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# type of files, VCF or GEN\n",
    "filetype=\"VCF\"\n",
    "\n",
    "# write snp log file or not \n",
    "snp_log= True\n",
    "## Setting parameters\n",
    "gwas_id=0 # column of SNP ID\n",
    "gwas_p=1     # column of P value\n",
    "gwas_or=2    # column of odds ratio\n",
    "gwas_a1=3    # column of a1 in the GWAS\n",
    "gwas_a2=4\n",
    "gwas_a1f=5  # column index of maf in the GWAS\n",
    "\n",
    "# defin column number for contents in genfile\n",
    "if filetype.lower()==\"vcf\":\n",
    "    chrom_id=0\n",
    "    bp_id=1\n",
    "    geno_id= 2 # column number with rsID\n",
    "    geno_start=9 # column number of the 1st genotype, in the raw vcf files, after separated by the delimiter of choice\n",
    "    geno_a1 = 3 # column number that contains the reference allele\n",
    "    GENO_delim= \"\\t\"\n",
    "    \n",
    "elif filetype.lower()==\"gen\":\n",
    "    chrom_id=0\n",
    "    geno_id = 1\n",
    "    bp_id=2\n",
    "    geno_start=5\n",
    "    geno_a1=3\n",
    "    GENO_delim= \" \"\n",
    "\n",
    "\n",
    "# List of thresholds:\n",
    "thresholds=[0.01, 0.05,0.1,0.2]\n",
    "threshold_seq=[0.001 ,0.01,0.002]\n",
    "            \n",
    "step=0.01  # default step size\n",
    "threshold_interval=[]\n",
    "if threshold_seq is not None:\n",
    "    if len(threshold_seq)==3:\n",
    "        lower=min(threshold_seq[0:2])\n",
    "        upper=max(threshold_seq[0:2])\n",
    "        step=threshold_seq[2]\n",
    "        threshold_interval=np.arange(lower, upper+step, step).tolist()\n",
    "    else:\n",
    "        raise(\"Invalid input for threshold sequence parameters\")\n",
    "        logger.error(\"Invalid input for threshold sequence parameters\")\n",
    "        \n",
    "thresholds=thresholds+threshold_interval\n",
    "# file delimiters:\n",
    "GWAS_delim=\"\\t\"\n",
    "\n",
    "# file names:\n",
    "#home=\"/Volumes/mavan/Genotyping_161114/MAVAN_imputed_161121/KIDS_info03/\"  #define homefolder path\n",
    "\n",
    "# Name of GWAS file\n",
    "gwasFiles=\"../TestGWAS.txt\"\n",
    "GWAS_has_header=True\n",
    "\n",
    "# programme parameter\n",
    "log_or=False  # sepcify whether you want to log your odds ratios\n",
    "check_ref=True # if you know that there are mismatch between the top strand in the genotypes and that of the GWAS, set True. Not checking the reference allele will improve the speed\n",
    "use_maf=True   # whether to use MAF to check reference allele\n",
    "\n",
    "# sample file path and name\n",
    "sampleFilePath=\"../KIDS.sample\" # include the full/relative path and name of the sample file\n",
    "sampleFileDelim=\" \"  # sample File Delimiter\n",
    "sampleFileID=[0]   # which column in the sample file has the ID\n",
    "sample_skip=2  # how many lines to skip so that the sample names can be matched to the genotypes 1-to-1, taking into account the header of the sample file\n",
    "##output file information\n",
    "\n",
    "outputPath=\"../TestResult\"\n",
    "\n",
    "# Sepcify whether to check for duplicate SNPs\n",
    "checkDup=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# get the name of the genotype files\n",
    "genoFileNamePattern=\"../Test.vcf\"\n",
    "if \"file:/\" in genoFileNamePattern:\n",
    "    genoFileNamePattern=re.sub(\"file://\", \"\", genoFileNamePattern)\n",
    "\n",
    "\n",
    "# get the whole list of the file names\n",
    "genoFileNames=glob.glob(genoFileNamePattern)\n",
    "\n",
    "# parameter for phenotype regression\n",
    "pheno_file=None\n",
    "#pheno_columns=results.pheno_columns\n",
    "#pheno_delim=results.pheno_delim\n",
    "#pheno_no_header=results.pheno_no_header\n",
    "#covar_columns=results.covar_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\" configure logging control \"\"\"\n",
    "\n",
    "logger = logging.getLogger(\"Test\")\n",
    "logger.setLevel(logging.DEBUG)\n",
    "# create file handler which logs even debug messages\n",
    "fh = logging.FileHandler(outputPath+\".log\")\n",
    "fh.setLevel(logging.DEBUG)\n",
    "# create console handler with a higher log level\n",
    "ch = logging.StreamHandler()\n",
    "ch.setLevel(logging.INFO)\n",
    "# create formatter and add it to the handlers\n",
    "formatter1 = logging.Formatter('%(asctime)s %(levelname)s : %(message)s')\n",
    "formatter2 = logging.Formatter('%(asctime)s %(levelname)s : %(message)s')\n",
    "\n",
    "ch.setFormatter(formatter1)\n",
    "fh.setFormatter(formatter2)\n",
    "# add the handlers to the logger\n",
    "logger.addHandler(fh)\n",
    "logger.addHandler(ch)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Reading Files\n",
      "Using these genoytpe files: \n",
      "../Test.vcf\n",
      "total of 1 files\n"
     ]
    }
   ],
   "source": [
    "##  start spark context\n",
    "import pyspark\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark import SparkConf, SparkContext\n",
    "from pyspark.sql import Row\n",
    "APP_NAME=\"PRS\"\n",
    "\n",
    "spark=SparkSession.builder.appName(APP_NAME).getOrCreate()\n",
    "\n",
    "# if using spark < 2.0.0, use the pyspark module to make Spark context\n",
    "# conf = pyspark.SparkConf().setAppName(APP_NAME).set()#.set(\"spark.serializer\", \"org.apache.spark.serializer.KryoSerializer\")\n",
    "\n",
    "sc  = spark.sparkContext\n",
    "\n",
    "#sc = spark.sparkContext\n",
    "sc.setLogLevel(\"WARN\")\n",
    "log4jLogger = sc._jvm.org.apache.log4j\n",
    "LOGGER = log4jLogger.LogManager.getLogger(__name__)\n",
    "print(\"Start Reading Files\")\n",
    "print(\"Using these genoytpe files: \")\n",
    "\n",
    "for filename in genoFileNames[:min(24, len(genoFileNames))]:\n",
    "    print(filename)\n",
    "if len(genoFileNames)>23:\n",
    "    print(\"and more...\")\n",
    "\n",
    "print(\"total of {} files\".format(str(len(genoFileNames))))\n",
    "# 1. Load files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using the GWAS file: ../TestGWAS.txt\n",
      "Showing top 5 rows of GWAS file\n",
      "+----------+-------+-------------------+---+---+---------+\n",
      "|       SNP|      P|               BETA| A1| A2|    CEUAF|\n",
      "+----------+-------+-------------------+---+---+---------+\n",
      "| rs3131967|0.06683| 0.0322157032979816|  T|  C|        .|\n",
      "|rs12562034| 0.8489|-0.0030070181092943|  A|  G|0.0925926|\n",
      "|rs12124819|0.05396| 0.0342272607705507|  A|  G|        .|\n",
      "| rs4970383|0.05491| 0.0277572046905535|  A|  C| 0.201835|\n",
      "| rs1806509|0.00927|-0.0338113190438628|  A|  C| 0.600917|\n",
      "+----------+-------+-------------------+---+---+---------+\n",
      "only showing top 5 rows\n",
      "\n",
      "System recognizes the following information in the GWAS :\n",
      "SNP ID : Column 0\n",
      "P-values : Column 1\n",
      "Effect size : Column 2\n",
      "Allele A1 : Column 3\n",
      "Allele A2 : Column 4\n",
      "Allele Frequencies : Column 5\n"
     ]
    }
   ],
   "source": [
    "# read the raw data\n",
    "genodata=sc.textFile(genoFileNamePattern)\n",
    "#print(\"Using the GWAS file: {}\".format(ntpath.basename(gwasFiles)))\n",
    "print(\"Using the GWAS file: {}\".format(gwasFiles))\n",
    "gwastable=spark.read.option(\"header\",GWAS_has_header).option(\"delimiter\", \"\\t\").csv(gwasFiles).cache()\n",
    "print(\"Showing top 5 rows of GWAS file\")\n",
    "gwastable.show(5)\n",
    "\n",
    "print(\"System recognizes the following information in the GWAS :\")\n",
    "print(\"SNP ID : Column {}\".format(gwas_id))\n",
    "print(\"P-values : Column {}\".format(gwas_p))\n",
    "print(\"Effect size : Column {}\".format(gwas_or))\n",
    "print(\"Allele A1 : Column {}\".format(gwas_a1))\n",
    "print(\"Allele A2 : Column {}\".format(gwas_a1+1))\n",
    "if use_maf:\n",
    "    print(\"Allele Frequencies : Column {}\".format(gwas_a1f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2017-03-13 10:19:53,261 INFO : Genotype data format : .VCF \n"
     ]
    }
   ],
   "source": [
    "# filter the genotype to contain only the SNPs less than the maximum p value threshold in the GWAS\n",
    "# Add a function that only keep the SNPs that match in chromosome and BP in geno and gwas\n",
    "\n",
    "maxThreshold=max(thresholds)  # maximum p value\n",
    "gwasOddsMapMax=filterGWASByP_DF(GWASdf=gwastable, pcolumn=gwas_p, idcolumn=gwas_id, oddscolumn=gwas_or, pHigh=maxThreshold, logOdds=log_or)\n",
    "gwasOddsMapMaxCA=sc.broadcast(gwasOddsMapMax).value  # Broadcast the map\n",
    "\n",
    "# ### 2. Initial processing\n",
    "# at this step, the genotypes are already filtered to keep only the ones in 'gwasOddsMapMax'\n",
    "bpMap={\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
    "tic=time.time()\n",
    "\n",
    "\n",
    "if filetype.lower()==\"vcf\":\n",
    "    logger.info(\"Genotype data format : .VCF \")\n",
    "\n",
    "    # Change to the format [snpid, A1, A2, *genotypelist]\n",
    "    genointermediate=genodata.filter(lambda line: (\"#\" not in line)).map(lambda line: line.split(GENO_delim)).filter(lambda line: line[geno_id] in gwasOddsMapMaxCA).map(lambda line:([line[x] for x in [geno_id, geno_a1, geno_a1+1]],[chunk.strip('\"').split(\":\")[3] for chunk in line[geno_start::]]))\\\n",
    "    .mapValues(lambda line:[float(x) for x in \",\".join(line).split(\",\")])  \n",
    "    \n",
    "elif filetype.lower() == \"gen\":\n",
    "    logger.info(\"Genotype data format : .GEN\")\n",
    "    # Change to the format [snpid, A1, A2, *genotypelist]\n",
    "    genointermediate=genodata\\\n",
    "    .filter(lambda line: line.split(GENO_delim)[geno_id] in gwasOddsMapMaxCA)\\\n",
    "    .map(lambda line: ([line.split(GENO_delim)[x] for x in [geno_id,geno_a1, geno_a1+1]], [float(x) for x in line.split(GENO_delim)[geno_start::]]))\n",
    "    \n",
    "\n",
    "# Change to the format [snpid, *genotypelist]\n",
    "genotable=genointermediate.map(lambda line: (line[0][0], line[1]))  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[u'##fileformat=VCFv4.2']"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genodata.first().split(\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(u'rs2251843',\n",
       " [0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.95,\n",
       "  0.05,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.65,\n",
       "  0.35,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.05,\n",
       "  0.95,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.95,\n",
       "  0.05,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.95,\n",
       "  0.05,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0,\n",
       "  0.0,\n",
       "  1.0,\n",
       "  0.0])"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotable.first()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2017-03-13 10:19:58,226 INFO : Determining strand alignment, using MAF\n",
      "2017-03-13 10:20:03,892 INFO : Generating genotype dosage while taking into account difference in strand alignment\n",
      "2017-03-13 10:20:04,200 INFO : Dosage generated in 10.9 seconds\n",
      "2017-03-13 10:20:04,496 INFO : Detected 260 samples in genotype data\n"
     ]
    }
   ],
   "source": [
    "if check_ref:\n",
    "    if use_maf:\n",
    "        logger.info(\"Determining strand alignment, using MAF\")\n",
    "        genoA1f=genointermediate.map(lambda line: (line[0]+[getA1f(line[1])])).toDF([\"Snpid_geno\", \"GenoA1\", \"GenoA2\", \"GenoA1f\"])\n",
    "        gwasA1f=gwastable.rdd.map(lambda line:(line[gwas_id], line[gwas_a1], line[gwas_a2], line[gwas_a1f])).toDF([\"Snpid_gwas\", \"GwasA1\", \"GwasA2\", \"GwasA1f\" ])\n",
    "        checktable=genoA1f.join(gwasA1f, genoA1f[\"Snpid_geno\"]==gwasA1f[\"Snpid_gwas\"], \"inner\").cache()\n",
    "        if checkDup:\n",
    "            logger.info(\"Searching and removing duplicated SNPs\")\n",
    "            flagList = checktable.rdd.map(lambda line: checkAlignmentDF(line, bpMap)).collect()\n",
    "            flagMap = rmDup(flagList)\n",
    "        else:\n",
    "            flagMap = checktable.rdd.map(lambda line: checkAlignmentDF(line, bpMap)).collectAsMap()\n",
    "    else:\n",
    "        logger.info(\"Determining strand alignment, without using MAF. SNPs with Alleles that are reverse compliments will be discarded\")\n",
    "        genoalleles=genotable.map(lambda line: (line[0])).toDF([\"Snpid_geno\", \"GenoA1\", \"GenoA2\"])\n",
    "        gwasalleles=gwastable.rdd.map(lambda line:(line[gwas_id], line[gwas_a1], line[gwas_a2])).toDF([\"Snpid_gwas\", \"GwasA1\", \"GwasA2\"])\n",
    "        checktable=genoalleles.join(gwasalleles, genoalleles[\"Snpid_geno\"]==gwasalleles[\"Snpid_gwas\"], \"inner\").cache()\n",
    "\n",
    "        if checkDup:\n",
    "            logger.info(\"Searching and removing duplicated SNPs\")\n",
    "            flagList = checktable.rdd.map(lambda line: checkAlignmentDFnoMAF(line, bpMap)).collect()\n",
    "            flagMap = rmDup(flagList)\n",
    "        else:\n",
    "            flagMap = checktable.rdd.map(lambda line: checkAlignmentDFnoMAF(line, bpMap)).collectAsMap()\n",
    "\n",
    "    logger.info(\"Generating genotype dosage while taking into account difference in strand alignment\")\n",
    "    flagMap=sc.broadcast(flagMap).value\n",
    "    genotypeMax=genotable.filter(lambda line: line[0] in flagMap and flagMap[line[0]]!=\"discard\" ).map(lambda line: makeGenotypeCheckRef(line, checkMap=flagMap)).cache()\n",
    "\n",
    "else:\n",
    "    logger.info(\"Generating genotype dosage without checking allele alignments\")\n",
    "    genotypeMax=genotable.mapValues(lambda line: makeGenotype(line)).cache()\n",
    "    flagMap=False\n",
    "    if checkDup:\n",
    "        logger.info(\"Searching and removing duplicated SNPs\")\n",
    "        genotypeCount=genotypeMax.map(lambda line: (line[0], 1)).reduceByKey(lambda a,b: a+b).filter(lambda line: line[1]==1).collectAsMap()\n",
    "        genotypeMax=genotypeMax.filter(lambda line: line[0] in genotypeCount)\n",
    "\n",
    "logger.info(\"Dosage generated in {:.1f} seconds\".format(time.time()-tic) )\n",
    "samplesize=int(len(genotypeMax.first()[1]))\n",
    "logger.info(\"Detected {} samples in genotype data\" .format(str(samplesize)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[((u'rs12611593', 0.009000000000000001), (u'rs12611593', 0.008991)),\n",
       " ((u'rs12120390', 0.009000000000000001), (u'rs12120390', 0.008995)),\n",
       " ((u'rs7461006', 0.009000000000000001), (u'rs7461006', 0.008997)),\n",
       " ((u'rs2034843', 0.009000000000000001), (u'rs2034843', 0.008997)),\n",
       " ((u'rs13111399', 0.01), (u'rs13111399', 0.009007)),\n",
       " ((u'rs7106686', 0.01), (u'rs7106686', 0.009009)),\n",
       " ((u'rs1349312', 0.01), (u'rs1349312', 0.009013)),\n",
       " ((u'rs16921167', 0.01), (u'rs16921167', 0.009015)),\n",
       " ((u'rs11605824', 0.01), (u'rs11605824', 0.009017)),\n",
       " ((u'rs7290176', 0.01), (u'rs7290176', 0.009019))]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## use the thresholds as bins, put each snp in the corresponding bins\n",
    "\n",
    "gwasP=gwastable.rdd.filter(lambda line: float(line[gwas_p])< maxThreshold).map(lambda line: (line[gwas_id], float(line[gwas_p]))).collect()\n",
    "\n",
    "def binTuple(snpwithP, thresholdList):\n",
    "  results=[]\n",
    "  snpwithPsorted=sorted(snpwithP,key=lambda x: x[1])\n",
    "  thresholdSorted=sorted(thresholdList)\n",
    "  thresholdidx=0\n",
    "  for snp, p in snpwithPsorted:\n",
    "\n",
    "    if p>thresholdSorted[thresholdidx]:\n",
    "      thresholdidx+=1    \n",
    "    results.append((snp,thresholdSorted[thresholdidx]))\n",
    "  return results\n",
    "  \n",
    "snpBin=binTuple(gwasP, thresholds)\n",
    "\n",
    "snpBinRDD=sc.parallelize(snpBin)\n",
    "\n",
    "genotypeMaxRanked=snpBinRDD.join(genotypeMax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#genoa1f.map(lambda line:\"\\t\".join([line[0], \"\\t\".join(line[1]), str(line[2])])).saveAsTextFile(\"../MOMS_info03_maf\")\n",
    "\n",
    "# Calculate PRS at the sepcified thresholds\n",
    "if flagMap:\n",
    "  genocalltable=genotable.filter(lambda line: line[0] in flagMap and flagMap[line[0]]!=\"discard\" ).mapValues(lambda geno: getCall(geno)).cache()\n",
    "else:\n",
    "  genocalltable=genotable.mapValues(lambda geno: getCall(geno))\n",
    "\n",
    "assert len(genocalltable.first()[1])==samplesize, \"Bug found, size of genotype and call table differ\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(u'rs6870608', (0.05, [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]))\n"
     ]
    }
   ],
   "source": [
    "\n",
    "genocalltableRanked=snpBinRDD.join(genocalltable)\n",
    "\n",
    "## multiply each call by the odds\n",
    "## sum up the score, and the calls, within each rank\n",
    "\n",
    "\n",
    "def calcIntervals(genotypeRDDRanked, gwasOddsMap, calltableRanked, logsnpON, logger=logger):\n",
    "  logger.info(\"Calculating scores in each bin\")\n",
    "  genotypeRDDMultipled=genotypeRDDRanked.map(lambda line: (line[1][0], [x*gwasOddsMap[line[0]] for x in line[1][1]]))\n",
    "  intervalScoreRDD=genotypeRDDMultipled.reduceByKey(lambda snp1, snp2: map(add, snp1, snp2))\n",
    "  intervalScores=intervalScoreRDD.collect()\n",
    "  \n",
    "  logger.info(\"Calculating calls in each bin\")\n",
    "  intervalCallsRDD=calltableRanked.map(lambda line:line[1]).reduceByKey(lambda snp1, snp2: map(add, snp1, snp2))\n",
    "  \n",
    "  intervalCalls=intervalCallsRDD.collect()\n",
    "\n",
    "  logger.info(\"Generating snp list in each bin\")\n",
    "  \n",
    "  snpLists=False\n",
    "  if logsnpON:\n",
    "    snpLists=calltableRanked.map(lambda line:(line[1][0], line[0])).groupByKey().map(lambda line: (line[0], list(line[1]))).collect()\n",
    "    \n",
    "  return intervalScores, intervalCalls, snpLists\n",
    "\n",
    "scoresBin, callsBin, snpBin=calcIntervals(genotypeMaxRanked, gwasOddsMapMaxCA, genocalltableRanked, snp_log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Take the sum of scores, calls and snplist in each bin and gather them\n",
    "def gatherScores(binScores,binCalls,binSNPs, thresholdList, logger=logger):\n",
    "  prsResults={}\n",
    "  snpNames={}\n",
    "  binScoresSorted=sorted(binScores)\n",
    "  binCallsSorted=sorted(binCalls)\n",
    "  binSNPsSorted=sorted(binSNPs)\n",
    "  logger.info(\"Start gathering scores from each bin\")\n",
    "  binThresholds=[x[0] for x in binScoresSorted]\n",
    "  for x in binThresholds:\n",
    "    if x not in thresholdList:\n",
    "      logger.info(\"No SNPs exist at threshold {}\".format(x))\n",
    "  \n",
    "  assert binThresholds==[x[0] for x in binCallsSorted], \"Error, scores and calls have different bins\"\n",
    "  assert binThresholds==[x[0] for x in binSNPsSorted], \"Error, scores and SNP list have different bins\"\n",
    "  \n",
    "  binScoresSortedvalues=[x[1] for x in binScoresSorted]\n",
    "  binCallsSortedvalues=[x[1] for x in binCallsSorted]\n",
    "  binSnpsSortedvalues=[x[1] for x in binSNPsSorted]\n",
    "  totalNumbers=len(binScores)\n",
    "  for i in range(len(binScoresSorted)):\n",
    "    \n",
    "    threshold=binThresholds[i]\n",
    "    scores=[sum(x) for x in zip(*binScoresSortedvalues[:(i+1)])]\n",
    "    calls=[sum(x) for x in zip(*binCallsSortedvalues[:(i+1)])]\n",
    "    normalizedScores=[score/call for score, call in zip(scores, calls)]\n",
    "    \n",
    "    prsResults[threshold]=[calls,normalizedScores]\n",
    "    combinedSNPs=reduce(lambda x,y: x+y, binSnpsSortedvalues[:(i+1)])\n",
    "    snpNames[threshold]=combinedSNPs\n",
    "    print(\"Processed {} / {} scores\".format(i+1, totalNumbers))\n",
    "    sys.stdout.flush()\n",
    "\n",
    "  return prsResults, snpNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[10, 9, 6, 12]"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=[2,3,2,4]\n",
    "b=[4,3,2,4]\n",
    "c=[4,3,2,4]\n",
    "reduce(lambda x,y:  map(add, x,y),[a,b,c])\n",
    "#[score/call for score, call in zip(a[1], b[1])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2017-03-13 11:58:30,963 INFO : Start gathering scores from each bin\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 1 / 9 scores\n",
      "Processed 2 / 9 scores\n",
      "Processed 3 / 9 scores\n",
      "Processed 4 / 9 scores\n",
      "Processed 5 / 9 scores\n",
      "Processed 6 / 9 scores\n",
      "Processed 7 / 9 scores\n",
      "Processed 8 / 9 scores\n",
      "Processed 9 / 9 scores\n"
     ]
    }
   ],
   "source": [
    "prsDict, snpids=gatherScores(scoresBin, callsBin, snpBin, thresholds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2017-03-13 11:58:53,670 INFO : Successfully output log to ../TestResult.snplog\n"
     ]
    }
   ],
   "source": [
    "# log which SNPs are used in PRS\n",
    "if snp_log:\n",
    "    if flagMap:\n",
    "        logoutput=writeSNPlog(snpids, outputPath, logger, flagMap=flagMap)\n",
    "    else:\n",
    "        logoutput=writeSNPlog(snpids, outputPath, logger)\n",
    "\n",
    "# generate labels for samples\n",
    "#if filetype.lower()==\"vcf\":\n",
    "    #subjNames=genodata.filter(lambda line: \"#CHROM\" in line).map(lambda line: line.split(GENO_delim)[9::]).collect()[0]\n",
    "    #output=writePRS(prsDict,  outputPath, samplenames=subjNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2017-03-13 11:59:41,406 INFO : Collected 260 sample labels\n",
      "2017-03-13 11:59:41,410 INFO : Successfully wrote scores to ../TestResult.score\n"
     ]
    }
   ],
   "source": [
    "if sampleFilePath!=\"NOSAMPLE\":\n",
    "    # get sample name from the provided sample file\n",
    "    subjNames=getSampleNames(sampleFilePath,sampleFileDelim,sampleFileID, skip=sample_skip)\n",
    "\n",
    "    output=writePRS(prsDict,  outputPath, logger, samplenames=subjNames)\n",
    "else:\n",
    "    output=writePRS(prsDict,  outputPath,logger=logger, samplenames=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'totalstart' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-299-c83378d05a25>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;31m#sc.stop()\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mseconds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mtotalstart\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0mm\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdivmod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseconds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m60\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdivmod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mm\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m60\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'totalstart' is not defined"
     ]
    }
   ],
   "source": [
    "if pheno_file is not None:\n",
    "    phenotypes, thresholds, r2All, pAll=regression(prsDict,pheno_file, pheno_delim, pheno_columns, pheno_no_header, covarColumns=covar_columns, outputName=outputPath, logger=logger)\n",
    "\n",
    "    r_square_plots(phenotypes,r2All,pAll, thresholds, outputName=outputPath, width = 3,bar_width = step)\n",
    "\n",
    "#sc.stop()\n",
    "seconds=time.time()-totalstart\n",
    "m, s = divmod(seconds, 60)\n",
    "h, m = divmod(m, 60)\n",
    "logger.info(\"Total Calculation Time : {:d} hrs {:02d} min {:02d} sec\".format(int(h), int(m), int(round(s))))"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
